wcfsetup/install/files/lib/acp/form/LanguageAddForm.class.php

```diff
@@ -103,149 +83,107 @@ public function readParameters()
-        if (!isset($this->locales[$this->locale])) {
-            throw new UserInputException('locale');
-        }
-    }
+        $this->languages = LanguageFactory::getInstance()->getLanguages();
 
-    /**
-     * Validates the language code.
-     */
-    protected function validateLanguageCode()
-    {
-        if (empty($this->languageCode)) {
-            throw new UserInputException('languageCode');
-        }
-        if (LanguageFactory::getInstance()->getLanguageByCode($this->languageCode)) {
-            throw new UserInputException('languageCode', 'notUnique');
-        }
+        $locales = [
+            '' => WCF::getLanguage()->get('wcf.global.noSelection'),
+            ...$this->locales
+        ];
+
+        $this->form->appendChildren([
+            FormContainer::create('data')
+                ->appendChildren([
+                    TextFormField::create('languageName')
+                        ->label('wcf.global.name')
+                        ->description('wcf.acp.language.name.description')
+                        ->maximumLength(255)
+                        ->required(),
+                    TextFormField::create('languageCode')
+                        ->label('wcf.acp.language.code')
+                        ->description('wcf.acp.language.code.description')
+                        ->maximumLength(20)
+                        ->required()
+                        ->addValidator(new FormFieldValidator('unique', function (TextFormField $formField) {
+                            if ($formField->getValidationErrors() !== []) {
+                                return;
+                            }
+
+                            if (
+                                $this->formObject instanceof Language
+                                && \mb_strtolower($this->formObject->languageCode) === $formField->getValue()
+                            ) {
+                                return;
+                            }
+
+                            if (LanguageFactory::getInstance()->getLanguageByCode($formField->getValue())) {
+                                $formField->addValidationError(new FormFieldValidationError(
+                                    'notUnique',
+                                    'wcf.acp.language.add.languageCode.error.notUnique'
+                                ));
+                            }
+                        })),
+                    TextFormField::create('countryCode')
+                        ->label('wcf.acp.language.countryCode')
+                        ->description('wcf.acp.language.countryCode.description')
+                        ->maximumLength(10)
+                        ->required(),
+                    SingleSelectionFormField::create('locale')
+                        ->label('wcf.acp.language.locale')
+                        ->description('wcf.acp.language.locale.description')
+                        ->options($locales)
+                        ->required(),
```

Comment:
I wonder if we should validate the `locale` against the `languageCode` to prevent dumb combinations like “de” with “en_US” locale.

@dtdesign What do you think?

Reply:
Yes, the `languageCode` and the `locale` must match, because the locale is just a more specific version of it.

Reply:
In that case we should also validate `languageCode` against ISO 639-1 and `countryCode` against ISO 3166-1. To be consequent, we could remove both fields (if a locale is selected) since they are defined by the locale.

Reply:
> To be consequent, we could remove both fields (if a locale is selected) since they are defined by the locale.

No. `languageCode` is still referenced a lot of times in both our and 3rd party code, including the language XML files. We could derive the language code from the locale though, but this is a journey for another day.

The `countryCode` on the other hand is used to pick which flag to show. It’s a bit awkward for English since we use American English, but use UK’s flag. Maybe one day we can get rid of the country flags, because it just makes things quite difficult without adding much benefit.

Reply:
What kind of validation would you expect? `^$languageCode_*` or `$languageCode === \explode('_', $locale)[0];`?
I don't see any proper way to do this without making the language and country code fields completely useless.

Reply:
The `explode()` variant is fine. With the regular expression you possibly need to mess around with escaping.

> I don't see any proper way to do this without making the language and country code fields completely useless.

And yes, it's a little redundant for now, but this is something to incrementally improve once we have experience in the field. This form is mostly going to be used by power-users who should be able to deal with this.

The “locale” selection is based on PHP’s intl extension and thus ICU. While our non-scientific research within the Smalltalk thread in our forum shows that intl is generally available and the ICU version also tends to be reasonably up to date, we don't yet know if there’s any way to break those as a web hoster. By separating the `languageCode` from the `locale` one can provide the actual localization even if the language is not (yet) supported by one’s ICU version. This might be a non-issue in practice, but we’ll find that out. We generally tend to do this kind of incremental improvement nowadays, because it allows one to react on real-world feedback instead of making a bunch of assumptions. One, purely technical, example might be #5237.